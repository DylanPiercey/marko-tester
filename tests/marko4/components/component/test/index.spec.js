const { render, fixtures, createEvent } = require('marko-tester')('../index.marko', { withoutFixtures: true });

fixtures();

describe('When component is rendered with data', () => {
  let component;
  const cachedFixture = JSON.stringify(fixtures.default);
  const mockUtil = 'mockUtil';

  beforeEach(() => {
    jest.mock('../../../../common/components/non-marko', () => mockUtil);

    component = render(fixtures.default);
  });

  afterEach(() => {
    component.destroy();
  });

  describe('When component is re-rendered', () => {
    beforeEach(() => {
      component.forceUpdate();
    });

    it('should re-render the component', () => {
      expect(true);
    });
  });

  it('should not mutate the fixture', () => {
    expect(JSON.stringify(fixtures.default)).toBe(cachedFixture);
  });

  it('should mock relative module', () => {
    expect(component.util).toBe(mockUtil);
  });

  it('should not render other component', () => {
    expect(component.getEl('tbody').tagName).toBe('TBODY-COMPONENT');
  });

  describe('When user press ENTER on the button', () => {
    let preventDefault;

    beforeEach(() => {
      preventDefault = jest.fn();

      component.getEl('button').dispatchEvent(createEvent('keypress', { keyCode: 13, preventDefault }));
    });

    it('should change the state and re-render', () => {
      expect(component.state.hidden).toBe(true);
      expect(component.getEl('tbody')).not.toBe(undefined);
      component.update();
      expect(component.getEl('tbody')).toBe(undefined);
    });

    it('should prevent default behavior', () => {
      expect(preventDefault).toBeCalled();
    });

    describe('When component is updated', () => {
      beforeEach(() => {
        component.update();
      });

      describe('When document is being clicked', () => {
        beforeEach(() => {
          jest.spyOn(component, 'emit');
          document.dispatchEvent(createEvent('click'));
        });

        it('should not execute event handler', () => {
          expect(component.emit).not.toBeCalled();
        });
      });
    });
  });

  describe('When user clicks emit button', () => {
    let clickEvent;

    beforeEach(() => {
      clickEvent = createEvent('click');

      jest.spyOn(component, 'emit');
      component.getEl('emit-button').dispatchEvent(clickEvent);
    });

    it('should emit "hello" with event and element', () => {
      expect(component.emit).toBeCalledWith('hello', clickEvent, clickEvent.target);
    });
  });

  describe('When document is being clicked', () => {
    beforeEach(() => {
      jest.spyOn(component, 'emit');
      document.dispatchEvent(createEvent('click'));
    });

    it('should execute event handler', () => {
      expect(component.emit).toBeCalledWith('document');
    });
  });
});

describe('When component is rendered without data', () => {
  let component;
  const mockUtil = 'mockUtil2';

  beforeEach(() => {
    jest.mock('../../../../common/components/non-marko', () => mockUtil);

    component = render(fixtures.hidden);
  });

  afterEach(() => {
    component.destroy();
  });

  it('should mock relative module', () => {
    expect(component.util).toBe(mockUtil);
  });
});

describe('When component is rendered without fixture', () => {
  let component;

  beforeEach(() => {
    component = render();
  });

  afterEach(() => {
    component.destroy();
  });

  it('should not error', () => {
    expect(true);
  });
});
