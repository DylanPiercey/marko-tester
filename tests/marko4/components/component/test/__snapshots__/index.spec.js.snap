// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component with default fixture 1`] = `
Array [
  ,
  <div
    class="mock-component"
  >
    <h1>
      hello
    </h1>
    <button
      type="button"
    >
      Hello
    </button>
    <button
      type="button"
    />
    
    
    <tbody-component
      attribute="{\\"title\\":\\"hello\\",\\"i\\":0,\\"hello\\":\\"world\\"}"
    >
      <at_body>
        body is here
      </at_body>
    </tbody-component>
  </div>,
  ,
]
`;

exports[`should render component with hidden fixture 1`] = `
Array [
  ,
  <div
    class="mock-component"
  >
    <h1>
      
    </h1>
    <button
      type="button"
    >
      Hello
    </button>
    <button
      type="button"
    />
  </div>,
  ,
]
`;

exports[`should render component with undefined fixture 1`] = `
Array [
  ,
  <div
    class="mock-component"
  >
    <h1>
      
    </h1>
    <button
      type="button"
    >
      Hello
    </button>
    <button
      type="button"
    />
    
    
    <tbody-component
      attribute="{\\"hello\\":\\"world\\"}"
    >
      <at_body>
        body is here
      </at_body>
    </tbody-component>
  </div>,
  ,
]
`;
